NEW: Advanced jobs admin UI.

mfendeksilverstripe opened the pull request:
# Advanced jobs admin UI

Optional advanced UI for jobs admin. Needs to be enabled via configuration API.

## Problem
Finding specific jobs in the Jobs admin is a chore, developers often find themselves scrolling through multiple pages. Finding the right jobs is sometimes impossible as the filter criteria input is quite limited. Using Text fields for input is often tedious.

## Solution
Improved UI. More filters, better input fields, more useful information in the columns.

![job-admin](https://user-images.githubusercontent.com/26395487/84096739-c3f8e700-aa56-11ea-9682-4454d2de61c5.png)

## Notes

* split from [Feature batch](https://github.com/symbiote/silverstripe-queuedjobs/pull/290) as `Feature 1 - Improved Jobs admin UI`

## Related issues
https://github.com/symbiote/silverstripe-queuedjobs/issues/308

mfendeksilverstripe committed:
NEW: Advanced jobs admin UI.

sachajudd commented:
@silverstripeux

emteknetnz commented on src/DataObjects/QueuedJobDescriptor.php:
```diff
@@ -302,6 +302,19 @@ public function getJobTypeString()
+    /**
+     * @return string
+     */
+    public function getImplementationSummary()
```
The use of the word 'summary' is a little weird, perhaps rename it to `getTrimmedImplementation()` or similar?.  I'd also add example to show that its trimming off the front of the namespace

emteknetnz commented on README.md:
```diff
@@ -819,6 +819,17 @@ As a consequence, the work might end up being very fragmented and each chunk may
 
 Some projects do not mind this however, so this solution may still be quite suitable.
 
+## Jobs admin UI
```
```suggestion
## Enabling advanced admin UI
```

sachajudd replied:
How about 'Enabling extended admin UI' instead of 'Advanced'? Even though this is the readme, we try to avoid advanced wording coupled in UI contexts where possible.

emteknetnz commented on src/Controllers/QueuedJobsAdmin.php:
```diff
@@ -74,6 +81,16 @@ class QueuedJobsAdmin extends ModelAdmin
 
+    /**
+     * Enables advanced admin UI
+     * Note that this requires optional module
```
```suggestion
     * This requires the optional module
```

emteknetnz reviewed (changes requested):
Look great, you'll need to rebase/retarget this to `4` since it contains new API

Cheddam reviewed (changes requested):
Can you also comb through and add translation calls for all the new language strings?

mfendeksilverstripe commented:
@emteknetnz @Cheddam Rebased on `4`, pushed up new changes, please review.

emteknetnz commented:
@mfendeksilverstripe I tried testing this locally though it didn't seem to activate the filterable headers.  I added the yml and installed the suggested module and ran /dev/build?flush

It's possible that it's related to merge conflicts after merging https://github.com/symbiote/silverstripe-queuedjobs/pull/314 - could you rebase this and see if it fixes things up?

mfendeksilverstripe commented:
@emteknetnz I rebased the branch. Also, I tested it on my local and it seems to be working fine. Have you enabled the UI on your local? It's disabled by default so you need to add following config to your YAML:

```
Symbiote\QueuedJobs\Controllers\QueuedJobsAdmin:
  advanced_admin_ui: true
```

emteknetnz commented:
@mfendeksilverstripe yeah I've required `silverstripe-terraformers/gridfield-rich-filter-header` though it does not appear to make any difference on at the top-level view `/admin/queuedjobs/`

I have included the following config and run /dev/build?flush

```yaml
Symbiote\QueuedJobs\Controllers\QueuedJobsAdmin:
  advanced_admin_ui: true
```

Also looks like you'll need to do another rebase (sorry) after I merged another one of your queuedjob prs

mfendeksilverstripe commented:
@emteknetnz I have rebased the branch and tested this feature on a vanilla install. It seems to be working fine. Here is my project config

![Screen Shot 2020-09-25 at 8 48 28 AM](https://user-images.githubusercontent.com/26395487/94198886-1aa60600-ff0c-11ea-8226-0619f7edf13c.png)

Note that I replaced the queued jobs module folder with this branch manually.

I recommend checking the version of the rich filter header module and also the following setting:

![Screen Shot 2020-09-25 at 8 48 43 AM](https://user-images.githubusercontent.com/26395487/94198987-3c9f8880-ff0c-11ea-9271-cdaf7f165e30.png)

This setting should be automatically set by the reich filter header module.

madmatt commented:
^ @emteknetnz Did you happen to get anywhere with this? Looks like a great addition and I'd be keen to see it merged in (noting there's a new conflict already heh)

mfendeksilverstripe commented:
@madmatt I've rebased the branch so it should be easy to merge in now.

mfendeksilverstripe commented:
Hey @emteknetnz here is a step by step guide which I used to test this feature on a vanilla install. Hopefully, this helps :)

1 - `composer create-project silverstripe/installer vanilla ^4`
2 - add `.env` and `vagrant` files
3 - `vagrant up`
4 - `composer install`
5 - `dev/build` with flush
6 - add fork to the composer file
```
"repositories": {
    "silverstripe/queuedjobs": {
        "type": "git",
        "url": "https://github.com/silverstripe-terraformers/silverstripe-queuedjobs.git"
    }
},
```
7 - `composer require symbiote/silverstripe-queuedjobs`
8 - change the queued jobs module in your composer JSON to point to the fork:

`"symbiote/silverstripe-queuedjobs": "dev-feature/jobs-admin-ui as 4.x-dev",`

9 - `composer update symbiote/silverstripe-queuedjobs`
10 - `composer require silverstripe-terraformers/gridfield-rich-filter-header`
11 - Enable the advanced UI via static configuration

```
Symbiote\QueuedJobs\Controllers\QueuedJobsAdmin:
  advanced_admin_ui: true
```
12 - `dev/build` with flush

This is what I see when I navigate to the Queued jobs admin:

![Screen Shot 2021-10-18 at 2 02 33 PM](https://user-images.githubusercontent.com/26395487/137652348-e5b06f11-3809-44e1-8e2a-32c79ce66884.png)

mfendeksilverstripe commented:
@emteknetnz  I don't think my env file is related to this but here it is just for reference:

```
SS_ENVIRONMENT_TYPE="dev"

SS_DATABASE_CLASS="MySQLDatabase"
SS_DEFAULT_ADMIN_USERNAME="admin"
SS_DEFAULT_ADMIN_PASSWORD="password"
SS_DATABASE_SERVER="localhost"
SS_DATABASE_USERNAME="root"
SS_DATABASE_PASSWORD="root"
SS_DATABASE_NAME="SS_vanilla"

SS_DATABASE_TIMEZONE="+00:00"
SS_BASE_URL="http://vanilla.local/"
SS_SEND_ALL_EMAILS_TO=""
```

I don't have anything out of ordinary in the vagrant file either. I'll try to add a unit test which would validate that the UI is loading properly.

mfendeksilverstripe committed:
PR feedback.

mfendeksilverstripe commented:
Hey @emteknetnz I've added unit tests to showcase that the new UI is working as expected. I recommend running the test suite on your local setup and see if it passes.

emteknetnz reviewed (changes requested):
Apologies, looking again at this with fresh eyes I'm really not sure that we should be essentially baking an third party optional module into queuedjobs.  I feel like things should be implemented in a bit more of an "extendable" fashion.

Most core maintainers won't have merge access on the terraformers account.  It seems like if we're going to integrate this module to this degree we'd probably need everything on the Silverstripe account.  However that goes against the idea that functionality like this would be better maintained outside of the core team

It also seems like you could simply use the existing `$this->extend('updateEditForm', $form);` hook within the terraformers module to to do everything you need to ... does this seem correct?

I'm happy with linking to the module within docs.  Not so sure about `suggest` in composer.json, does that still work in composer 2?

mfendeksilverstripe committed:
Fixed a typo.

mfendeksilverstripe commented:
Thanks for clarification, I'll close this PR then.
